instruction-view/src/cpi.rs

```diff
@@ -0,0 +1,733 @@
+/// consuming CUs in the process. The [`invoke_signed_unchecked`] is an alternative
+/// to this function that have lower CU consumption since it does not perform
+/// any validation. This should only be used when the caller is sure that the borrow
+/// checker rules are followed.
+///
+/// Note that the maximum number of accounts that can be passed to a cross-program
+/// invocation is defined by the `MAX_CPI_ACCOUNTS` constant. Even if the slice
+/// of `AccountView`s has more accounts, only the number of accounts required by
+/// the `instruction` will be used. If the number of accounts required by the
+/// instruction is greater than [`MAX_CPI_ACCOUNTS`], this function will return a
+/// [`ProgramError::InvalidArgument`] error.
+///
+/// # Important
+///
+/// The accounts on the `account_views` slice must be in the same order as the
+/// `accounts` field of the `instruction`. When the instruction has duplicated
+/// accounts, it is necessary to pass a duplicated reference to the same account
+/// to maintain the 1:1 relationship between `account_views` and `accounts`.
+pub fn slice_invoke_signed(
+    instruction: &InstructionView,
+    account_views: &[&AccountView],
+    signers_seeds: &[Signer],
+) -> ProgramResult {
+    // Check that the stack allocated account storage `MAX_CPI_ACCOUNTS` is
+    // sufficient for the number of accounts expected by the instruction.
+    //
+    // The check for the slice of `AccountView`s not being less than the
+    // number of accounts expected by the instruction is done in
+    // `invoke_signed_with_bounds`.
+    if MAX_CPI_ACCOUNTS < instruction.accounts.len() {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    // SAFETY: The stack allocated account storage `MAX_CPI_ACCOUNTS` was validated.
+    unsafe {
+        inner_invoke_signed_with_bounds::<MAX_CPI_ACCOUNTS>(
+            instruction,
+            account_views,
+            signers_seeds,
+        )
+    }
+}
+
+/// Internal function to invoke a cross-program instruction with signatures
+/// from a slice of `AccountView`s performing borrow checking.
+///
+/// This function performs validation of the `account_views` slice to ensure that:
+///   1. It has at least as many accounts as the number of accounts expected by
+///      the instruction.
+///   2. The accounts match the expected accounts in the instruction, i.e., their
+///      `Address` matches the `address` in the `AccountView`.
+///   3. The borrow state of the accounts is compatible with the mutability of the
+///      accounts in the instruction.
+///
+/// # Safety
+///
+/// This function is unsafe because it does not check that the stack allocated account
+/// storage `MAX_ACCOUNTS` is sufficient for the number of accounts expected by the
+/// instruction. Using a value of `MAX_ACCOUNTS` that is less than the number of accounts
+/// expected by the instruction will result in undefined behavior.
```

Comment:
nit: To be honest, I don't know if this function is unsafe. This sounds more like a usability error to me. The invoked program should error if it doesn't receive a well-formed instruction, but it won't cause an issue like reading uninitialized memory. Am I understanding this correctly?